File: src/kio/workerbase.cpp
// SPDX-FileCopyrightText: 2024 Nicolas Fella <nicolas.fell@gmx.de>
// SPDX-License-Identifier: MPL-2.0

#include "cxx-kde-frameworks/workerbase.h"

#include <QUrl>

WorkerGlue::WorkerGlue(const QByteArray &protocol, const QByteArray &poolSocket, const QByteArray &appSocket, ::WorkerBase *worker)
    : KIO::WorkerBase(protocol, poolSocket, appSocket)
    , m_worker(worker)
{
}

KIO::WorkerResult WorkerGlue::get(const QUrl &url)
{
    return KIO::WorkerResult(*m_worker->get(url));
}

KIO::WorkerResult WorkerGlue::listDir(const QUrl &url)
{
    return KIO::WorkerResult(*m_worker->listDir(url));
}

KIO::WorkerResult WorkerGlue::stat(const QUrl &url)
{
    return KIO::WorkerResult(*m_worker->stat(url));
}

KIO::WorkerResult WorkerGlue::write(const QByteArray &data)
{
    return KIO::WorkerResult(*m_worker->write(data));
}

WorkerBase::WorkerBase(const QByteArray &protocol, const QByteArray &poolSocket, const QByteArray &appSocket)
    : m_glue(protocol, poolSocket, appSocket, this)
{
}

void WorkerBase::dispatchLoop()
{
    m_glue.dispatchLoop();
}

void WorkerBase::listEntry(const KIO::UDSEntry &entry)
{
    m_glue.listEntry(entry);
}

void WorkerBase::statEntry(const KIO::UDSEntry &entry)
{
    m_glue.statEntry(entry);
}

void WorkerBase::dataReq()
{
    m_glue.dataReq();
}

void WorkerBase::workerStatus(const QString &host, bool connected)
{
    m_glue.workerStatus(host, connected);
}

bool WorkerBase::canResume(KIO::filesize_t offset)
{
    return m_glue.canResume(offset);
}

void WorkerBase::data(const QByteArray &data)
{
    m_glue.data(data);
}

std::unique_ptr<WorkerResult> workerResultPass()
{
    return std::make_unique<KIO::WorkerResult>(KIO::WorkerResult::pass());
}

std::unique_ptr<WorkerResult> workerResultFail(int code, const QString &message)
{
    return std::make_unique<KIO::WorkerResult>(KIO::WorkerResult::fail(code, message));
}

File: src/kio/workerbase.rs
// SPDX-FileCopyrightText: 2024 Nicolas Fella <nicolas.fell@gmx.de>
// SPDX-License-Identifier: MPL-2.0

#[cxx_qt::bridge]
mod ffi {

    unsafe extern "C++" {
        include!("cxx-qt-lib/qbytearray.h");
        type QByteArray = cxx_qt_lib::QByteArray;

        include!("cxx-qt-lib/qstring.h");
        type QString = cxx_qt_lib::QString;

        include!("cxx-kde-frameworks/udsentry.h");
        type UDSEntry = crate::kio::udsentry::UDSEntry;
    }

    unsafe extern "C++" {
        include!("cxx-kde-frameworks/workerbase.h");

        type WorkerBase;

        type WorkerResult;

        pub fn data(self: Pin<&mut WorkerBase>, data: &QByteArray);

        pub fn listEntry(self: Pin<&mut WorkerBase>, entry: &UDSEntry);

        pub fn workerResultPass() -> UniquePtr<WorkerResult>;

        pub fn workerResultFail(code: i32, message: &QString) -> UniquePtr<WorkerResult>;

    }
}

use cxx::UniquePtr;
use cxx_qt_lib::QString;

impl WorkerResult {
    pub fn pass() -> UniquePtr<WorkerResult> {
        ffi::workerResultPass()
    }

    pub fn fail(code: i32, message: &QString) -> UniquePtr<WorkerResult> {
        ffi::workerResultFail(code, message)
    }
}

pub use ffi::WorkerBase;
pub use ffi::WorkerResult;
